### PR TITLE
Stefanozampini/small improvements

### PR DESCRIPTION
**Pull request by stefanozampini**
This branch contains a bugfix as discussed in https://github.com/mfem/mfem/issues/443 as well as few other small changes.

Except for the PETSc classes, the main change consists in promoting all coefficients of MFEM bilinear integrators from `private` to `protected`, so that they could be accessible from derived classes (I am working at the PDE constrained optimization code that makes uses of this https://github.com/stefanozampini/petscopt/)

New method SparseMatrix::Chop: compresses out the CSR format and remove zeros (I have used it for running some tests with LOR Elasticity) 

Also, I have fixed a problem in FindPoints for NC meshes

### CODE REVIEW

**Comment by stefanozampini** [comment]
@tzanio Do you think this branch can make it to `mfem-4.0`?

**Comment by tzanio** [comment]
> Do you think this branch can make it to mfem-4.0?

Possibly. @v-dobrev, what do you think?

**Comment by v-dobrev** [comment]
Yes, I think this PR should be reviewed and included in mfem-4.0.

**Comment by tzanio** [comment]
@stefanozampini, can you take a look while the Maxwell miniapp run fails on Travis?

**Comment by stefanozampini** [comment]
@tzanio I tried reproducing the environment used in the Travis machine, but I cannot reproduce the problem on my machine. The only difference is the METIS install (4.0 on Travis, 5.0 locally).
Since the error is caused by the last commit of this branch, I suspect the error is originated by a process having zero elements. `fe` should be null in this case, and https://github.com/mfem/mfem/pull/756/commits/0c1318dfd363afca4de3486501155fa85fbdac3a assumes a scalar FE in this case. Without this commit, the code should have segfaulted, but for some reasons it did not.

File: linalg/sparsemat.hpp [hunk 1]
```diff
@@ -320,6 +320,11 @@ class SparseMatrix : public AbstractSparseMatrix
    bool Finalized() const { return (A != NULL); }
    bool areColumnsSorted() const { return isSorted; }
 
+   /** @brief Remove entries smaller in absolute value than a given
+       tolerance @a tol. If @a fix_empty_rows is true, a zero value
+       is inserted in the diagonal entry (for square matrices only) */
+   void Chop(double tol, bool fix_empty_rows = false);
```
Comment (tzanio) [inline linalg/sparsemat.hpp]: A similar function in `HypreParMatrix` is called `Threshold`.

**Comment by tzanio** [comment]
Thank you @stefanozampini, do you want to mention any of the changes in`CHANGELOG`?

@jakubcerveny and @v-dobrev -- any objections to merging this?

File: mesh/mesh.cpp [hunk 2]
```diff
@@ -9492,6 +9492,27 @@ int Mesh::FindPoints(DenseMatrix &point_mat, Array<int>& elem_ids,
                }
             }
          }
+         // Try neighbours for non-conforming meshes
+         if (ncmesh)
+         {
+            Array<int> neigh;
+            int le = ncmesh->leaf_elements[e_idx[k]];
+            ncmesh->FindNeighbors(le,neigh);
```
Comment (jakubcerveny) [inline mesh/mesh.cpp]: For a single element, FindNeighbors() is unfortunately expensive (O(N))...

File: .gitignore [hunk 3]
```diff
@@ -183,3 +183,6 @@ miniapps/nurbs/Example1*
 # Unit test binary and outputs
 tests/unit/output_meshes
 tests/unit/unit_tests
+
+# VPATH builds
+build-*/*
```
Comment (v-dobrev) [inline .gitignore]: What's a VPATH build? Just curious.

File: linalg/petsc.hpp [hunk 4]
```diff
@@ -139,9 +141,12 @@ class PetscParVector : public Vector
    /// Returns the global number of rows
    PetscInt GlobalSize() const;
 
-   /// Conversion function to PETSc's Vec type
+   /// Typecasting to PETSc's Vec type
```
Comment (v-dobrev) [inline linalg/petsc.hpp]: The term "conversion function" comes from the c++ standard (12.3.2), see e.g. https://en.cppreference.com/w/cpp/language/cast_operator. No need to change it -- just FYI.

**Comment by tzanio** [comment]
@stefanozampini, we merged this for testing but please see the comments below and let us know early next week if you'd like to make any additional changes.

**Comment by stefanozampini** [comment]
@v-dobrev Just another name for out-of-source builds https://www.gnu.org/software/automake/manual/html_node/VPATH-Builds.html

**Comment by stefanozampini** [comment]
For the typecasting, just to use the same comments as for the other classes (PetscParMatrix, etc)

**Comment by stefanozampini** [comment]
I'm fine with this branch. @v-dobrev, thanks for the fixes to example.  If you wish to make any further changes, go ahead. 
@jakubcerveny If you have any faster alternative to the FindNeighbours routine, go ahead and push it.
For now, I just want to have points correctly located in the elements.

**Comment by v-dobrev** [comment]
> @v-dobrev Just another name for out-of-source builds https://www.gnu.org/software/automake/manual/html_node/VPATH-Builds.html

I see, thanks for the clarification.

> I'm fine with this branch. @v-dobrev, thanks for the fixes to example. If you wish to make any further changes, go ahead.

I'm not going to make any further changes - this is good to go from my perspective.

**Comment by jakubcerveny** [comment]
> @jakubcerveny If you have any faster alternative to the FindNeighbours routine, go ahead and push it.

It's on my TODO list, but I don't have a faster alternative at the moment, sorry.

**Pull request merged**